staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go

```diff
@@ -116,15 +110,42 @@ func WithPriorityAndFairness(
 				waitingMark.recordReadOnly(int(atomic.AddInt32(&atomicReadOnlyWaiting, delta)))
 			}
 		}
+		var resultCh chan interface{}
 		execute := func() {
 			noteExecutingDelta(1)
 			defer noteExecutingDelta(-1)
 			served = true
-			innerCtx := context.WithValue(ctx, priorityAndFairnessKey, classification)
-			innerReq := r.Clone(innerCtx)
+
+			innerCtx := ctx
+			innerReq := r
+
+			var watchInitializationSignal utilflowcontrol.InitializationSignal
+			if isWatchRequest {
+				watchInitializationSignal = newInitializationSignal()
+				innerCtx = utilflowcontrol.WithInitializationSignal(ctx, watchInitializationSignal)
+				innerReq = r.Clone(innerCtx)
+			}
 			setResponseHeaders(classification, w)
 
-			handler.ServeHTTP(w, innerReq)
+			if isWatchRequest {
+				resultCh = make(chan interface{})
+				go func() {
+					defer func() {
+						// We don't handle potential panics for backward compatibility.
+						resultCh <- recover()
+					}()
+
+					// Protect from the situations when request will not reach storage layer
+					// and the initialization signal will not be send.
+					defer watchInitializationSignal.Signal()
+
+					handler.ServeHTTP(w, innerReq)
```

Comment:
Readding the other comment from @tkashem 

with Watch included now, I think we might have a forever blocked goroutine.
StartRequest 1 function spins up a go routine 2 and it waits until the request is done using _ = <-ctx.Done(). If I am not mistaken the request context for watch is never canceled or does not have any deadline, this will cause the goroutine to block forever, I think.

We might need to do some plumbing so we have something similar to:

select {
case <-ctx.Done():
case <-watchInitializationSignal.Wait():
}

...


so I just did a quick scan of the filters and looks like for a Watch we do not apply any WithCancel or WithTimeout to the request context.
The only filter that sets a deadline to a request before the APF filter is WithRequestDeadline and it only applies it to non long running requests.

I also checked the net/http2 library, looks like just before the http handler is called, net/http creates a context with WithCancel

	ctx, cancelCtx := context.WithCancel(sc.baseCtx)
	st := &http2stream{
		sc:        sc,
		id:        id,
		state:     state,
		ctx:       ctx,
		cancelCtx: cancelCtx,
	}
``
and then creates a new `http.Request` with this context and then invokes the handler. So I think I may have been wrong when I said that `tx.Done() == nil` and we never create that go routine. (let me do some tests to verify that)

Reply:
I will write a one-time test to double check the `ctx.Done()` channel on a brand new HTTP request, but I think the goroutine will live as long the watch request lives, it should not hang forever, the `context` associated with the http2 stream of the request should get canceled when the watch ends.

Reply:
OK - I added a unit test to ensure that's the case.
This PR seem to be ready for re-review now.

Reply:
thanks for the unit test @wojtek-t, so now the goroutine will run as long as the watch request lasts. One optimization we can think of is that this goroutine can wait until the watch initialization is signaled as opposed to waiting for the watch request to complete (this can be addressed in a separate PR though)